### DISCUSSION TITLE
[spacy-nightly] Why does TextCatCNN average the transformer output features over the sequence?

### DISCUSSION THREAD

**Discussion by jricheimer** [explosion/spaCy#6501] [Help: Coding & Implementations] [answered]
I'm using spacy-nightly (v3.0.0rc2) for training a TextCat component with a transformer.

It looks like the default registered TextCatCNN takes the output features from the transformer over the whole sequence of tokens and uses a `reduce_mean` over the sequence dimension before applying a linear+softmax layer.

My understanding is that the standard way to do sequence classification with a transformer was to add the softmax layer on the output features only of the first `[CLS]` token. As with, for example, [`transformers.BertForSequenceClassification`](https://github.com/huggingface/transformers/blob/443f67e887a030d8254eba126e5f2cdb8b70eb63/src/transformers/models/bert/modeling_bert.py#L1381)  from HuggingFace.

Is there a reason TextCatCNN is doing things differently here?

**Comment by svlandeg** [comment, marked as answer]
Good question! 

You're right that a `reduce_mean` is applied as default `pooling` layer to obtain token vectors from the wordpiece vectors, and these token vectors are then fed into the output layers of the `TextCatCNN` directly. I could be mistaken - but I think the simple reason why we've done it like this, is because the `TextCatCNN` is agnostic of whether or not it's using a Transformer or some other [`Tok2Vec`](https://nightly.spacy.io/api/tok2vec) layer. This allows us to easily swap a "standard" spaCy `Tok2Vec` layer out for a `Transformer`, while still using the same output classification layers. And we wouldn't want to do summarization of the token vectors up-front before feeding it into the classification layers, because we want to be able to [share](https://nightly.spacy.io/usage/embeddings-transformers#embedding-layers) the same `Tok2Vec` or `Transformer` layer also for token-level tasks such as tagging. 

The nice thing is that with spaCy 3, you should be able to implement and register your own [model architectures](https://nightly.spacy.io/usage/layers-architectures), wire those into your `textcat`, and see how the alternative approach performs. I guess these benchmarks will also greatly depend on the type of data you're dealing with, but I'd be interested in hearing those results if you do decide to go that route!

**Comment by jricheimer** [comment]
Thanks @svlandeg for the explanation. I might follow your suggestion with the custom architecture, and I'll report back here with my observations if I do.

**Reply by ketantodi** [reply]
Hello @jricheimer  were you able to train the text classifier by using softmax on top of the [CLS] token output, or were you able to get the CLS token representation from the transformer model in spacy.

**Reply by jricheimer** [reply]
Hi @ketantodi , yes we did this by writing a custom TransformerListener, and it turned out that it improved the accuracy significantly on our dataset.

**Reply by ketantodi** [reply]
Hi @jricheimer thanks for replying back,  I will recheck the [documentation](https://spacy.io/usage/layers-architectures) for writing a custom TransformerListener, and will post back again if I bump into any issues.

**Reply by jricheimer** [reply]
@ketantodi There is more info that would probably be helpful to you [here](https://github.com/explosion/spaCy/issues/7139#issuecomment-782791948).
